### ISSUE TITLE
How to access the edge attributes

### ISSUE THREAD

**Issue by greenappletree** [rusty1s/pytorch_geometric#693]
## ❓ Questions & Help

Hi, 
PyG is a great library. Thanks for your work. 
We have a question that how to access the edge attributes in the message function, like node features x_j?

Thanks.

**Comment by rusty1s** [comment]
Just pass them to `propagate` and add them as arguments to the message function :)
```python
self.propagate(edge_index, x=x, edge_attr=edge_attr)
...
def message(self, x_j, edge_attr):
   ...
```

**Comment by greenappletree** [comment]
Thanks for your reply. 

Does the edge attribute (edge_attr) correspond to the nodes x_i and x_j ?

**Comment by rusty1s** [comment]
Yes, of course :)

**Comment by greenappletree** [comment]
Thank you very much!

**Comment by greenappletree** [comment]
Hi, 
I am confused when I run the example mnist_nn_conv.py.  
     https://github.com/rusty1s/pytorch_geometric/blob/00250118f4d250f95b24ed61ba34b75450bfacb0/torch_geometric/nn/conv/nn_conv.py#L83

I wonder that if the variable edge_attr and the variable pseudo are the same.

**Comment by rusty1s** [comment]
Yes, this formulation stems from the `MoNet` paper and refers to relative spatial information encoded in the edge features. We use them interchangeably in PyG.

**Comment by greenappletree** [comment]
Hi, 
NNConv.py
` 98     def forward(self, x, edge_index, edge_attr):
 99         """"""
100         """
101         print (x.size())
102         print (edge_index.size())
103         print (edge_attr.size())
104         print ('===')
105         """
106         x = x.unsqueeze(-1) if x.dim() == 1 else x
107         pseudo = edge_attr.unsqueeze(-1) if edge_attr.dim() == 1 else edge_attr
108         # print ('pseudo=', pseudo)
109         print ('x', x)
110         print ('edge_attr', edge_attr)
111         input()
112         return self.propagate(edge_index, x=x, edge_attr=edge_attr)
113
114     def message(self, x_i, x_j, edge_attr):
115         print ('x_i', x_i)
116         print ('x_j', x_j)
117         print ('edge_attr', edge_attr)
118         input()
119         # print (x_j.size())
120         # print (pseudo.size())
121
122         # input()
123         weight = self.nn(pseudo).view(-1, self.in_channels, self.out_channels)
124         # print (weight.size())
125         # input()
126         return torch.matmul(x_j.unsqueeze(1), weight).squeeze(1)
`

main.py
`  1
  2 import torch
  3 import torch.nn as nn
  4 import torch.nn.functional as F
  5 from nn_conv import NNConv
  6 from torch_geometric.data import Data
  7
  8
  9 edge_index = torch.tensor(
 10     [[1, 0, 1, 3, 3],
 11      [0, 1, 2, 0, 2]], dtype=torch.long
 12 )
 13
 14 x = torch.tensor([[0, 1], [1, 2], [2, 3], [3, 4]], dtype=torch.float)
 15
 16 edge_attr = torch.tensor(
 17     [[0.1], [0.1], [0.2], [0.4], [0.3]], dtype=torch.float
 18 )
 19
 20 data = Data(x=x, edge_index=edge_index, edge_attr=edge_attr)
 21
 22 print (data)
 23
 24
 25
 26 class Net(nn.Module):
 27     def __init__(self):
 28         super(Net, self).__init__()
 29         nn1 = nn.Sequential(nn.Linear(2, 25), nn.ReLU(), nn.Linear(25, 32))
 30         self.conv1 = NNConv(data.num_features, 32, nn1, aggr='mean')
 31
 32     def forward(self, data):
 33         data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
 34
 35
 36
 37 device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 38 model = Net().to(device)
 39 optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
 40
 41
 42 data = data.to(device)
 43 optimizer.zero_grad()
 44 model(data)
`
The corresponding outputs are presented as follows:
Data(edge_attr=[5, 1], edge_index=[2, 5], x=[4, 2])
x tensor([[0., 1.],
        [1., 2.],
        [2., 3.],
        [3., 4.]], device='cuda:0')
edge_attr tensor([[0.1000],
        [0.1000],
        [0.2000],
        [0.4000],
        [0.3000]], device='cuda:0')

x_i tensor([[0., 1.],
        [1., 2.],
        [2., 3.],
        [0., 1.],
        [2., 3.]], device='cuda:0')
x_j tensor([[1., 2.],
        [0., 1.],
        [1., 2.],
        [3., 4.],
        [3., 4.]], device='cuda:0')
edge_attr tensor([[0.1000],
        [0.1000],
        [0.2000],
        [0.4000],
        [0.3000]], device='cuda:0')

We are confused that the variable edge_attr in the message function is the same as that in the forward function.  How does the variable edge_attr work? Thanks.

**Comment by rusty1s** [comment]
It indeed stays the same. The `message` function maps each node information to source and target information respectively. There, you can also incorporate edge features to compute the final messages which gets then send along the edges.

**Comment by greenappletree** [comment]
Thanks for your reply!

**Comment by Chokerino** [comment]
> Just pass them to `propagate` and add them as arguments to the message function :)
> 
> ```python
> self.propagate(edge_index, x=x, edge_attr=edge_attr)
> ...
> def message(self, x_j, edge_attr):
>    ...
> ```

But this gives the entire attr list. Is there any way to get the attr for the edge from i to j?
